File: JULY-2025/10-07-2025/p3440.scala
object Solution {
  def maxFreeTime(eventTime: Int, startTime: Array[Int], endTime: Array[Int]): Int = {
    val n = startTime.length
    if (n == 0) return eventTime

    val gaps = Array.ofDim[Int](n + 1)
    gaps(0) = startTime(0)
    for (i <- 1 until n) {
      gaps(i) = startTime(i) - endTime(i - 1)
    }
    gaps(n) = eventTime - endTime(n - 1)

    val largestRight = Array.ofDim[Int](n + 1)
    for (i <- (0 until n).reverse) {
      largestRight(i) = math.max(largestRight(i + 1), gaps(i + 1))
    }

    var maxFree = 0
    var largestLeft = 0

    for (i <- 1 to n) {
      val duration = endTime(i - 1) - startTime(i - 1)
      val canFitLeft = largestLeft >= duration
      val canFitRight = largestRight(i) >= duration

      if (canFitLeft || canFitRight) {
        val merged = gaps(i - 1) + gaps(i) + duration
        maxFree = math.max(maxFree, merged)
      }

      val current = gaps(i - 1) + gaps(i)
      maxFree = math.max(maxFree, current)
      largestLeft = math.max(largestLeft, gaps(i - 1))
    }

    maxFree
  }
}

File: JULY-2025/10-07-2025/p3440.rb
def max_free_time(event_time, start_time, end_time)
  n = start_time.length
  return event_time if n == 0

  gaps = Array.new(n + 1, 0)
  gaps[0] = start_time[0]
  (1...n).each { |i| gaps[i] = start_time[i] - end_time[i - 1] }
  gaps[n] = event_time - end_time[n - 1]

  largest_right = Array.new(n + 1, 0)
  (n - 1).downto(0) do |i|
    largest_right[i] = [largest_right[i + 1], gaps[i + 1]].max
  end

  max_free = 0
  largest_left = 0

  (1..n).each do |i|
    duration = end_time[i - 1] - start_time[i - 1]
    can_fit_left = largest_left >= duration
    can_fit_right = largest_right[i] >= duration

    if can_fit_left || can_fit_right
      merged = gaps[i - 1] + gaps[i] + duration
      max_free = [max_free, merged].max
    end

    current = gaps[i - 1] + gaps[i]
    max_free = [max_free, current].max
    largest_left = [largest_left, gaps[i - 1]].max
  end

  max_free
end